pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -68,6 +69,12 @@ const (
 	// This field is deprecated. v1.Service.PublishNotReadyAddresses will replace it
 	// subsequent releases.  It will be removed no sooner than 1.13.
 	TolerateUnreadyEndpointsAnnotation = "service.alpha.kubernetes.io/tolerate-unready-endpoints"
+
+	// IsHeadlessServiceAnnotation is an annotation on the Endpoint denoting if its parent
+	// Service is Headless. This annotation can be used further by other controllers and
+	// kube-proxy to check if the Endpoint objects should be replicated when using Headless
+	// Services
+	IsHeadlessServiceAnnotation = "service.kubernetes.io/headless"
```

Comment:
@freehan @robscott PTAL at this wrt EndpointSlice, specifically let's agree on naming.

I think you were planning on `kubernetes.io/service` ? 

Because we have manual service-endpoint pairs, we can't really make this a positive test (has a VIP), so I think it has to stay in this polarity (has no VIP).  The question is whether this should be `service.kubernetes.io` or `endpoints.kubernetes.io` or `svc-ctrl.kubernetes.io` or `kube-proxy.kubernetes.io`?   I think I lean towards `service.kubernetes.io` but not strongly.

Thoughts?

Reply:
@thockin Agree! Actually I've just put something that came to my mind in this annotation name, with the hope that someone with far more experience (in this case, YOU) could put some more coherent annotation naming.

I do agree it could be something like ``service.kubernetes.io/has-no-vip`` as naming through kube-proxy or svc-ctrl could lead to something not directly related to what this does (as this annotation could be used not only by kube-proxy, etc etc).

Anyway I'll wait until the folks working into the EndpointSlice take a look into this to agree on naming :)

Reply:
We were planning on `kubernetes.io/service`, though that really doesn't look ideal after looking through what else is in use. Cloud load balancers are using a `kubernetes.io/service-name` annotation already which would match nicely with our EndpointSlice use case, so instead of creating a new annotation, it likely makes the most sense for us to stick with what's already in use. 

With that said, newer service related annotations seem to match the `service.kubernetes.io` scheme, including `service.kubernetes.io/service-proxy-name` and `service.kubernetes.io/load-balancer-cleanup`, so I'd agree with using something with that prefix here.

Reply:
An annotation should indicate who owns it's meaning (organizationally) and have enough context that someone could figure it out.

Pardon while I think out loud for a moment:

wrt EndpointSlice there's not a "backlink" per se, but there *IS* OwnerReference, which we should make SURE to set.  That allows someone to back-track from an EPS to the Service that owns it.  That is not selectable, so we need something in labels.  BUT the thing that we stick in labels doesn't have to be friendly or generic, it just has to be discoverable.   What `kubernetes.io/service` or `kubernetes.io/service-name` has going for it is that it is implicit and known a priori.  I don't need to write anything into Service to say "here's how you find my endpoints".  One MIGHT argue that we should use UUID or something, but I don't know that it adds any value.  So at the end of this, I think `kubernetes.io/service-name` is fine.  We really just need to document the relationship.

Back to this PR.

We already have a const `LabelServiceProxyName` which uses `service.kubernetes.io/service-proxy-name` so I think `service.kubernetes.io/headless` is a reasonable name for this.

the existing `service-proxy-name` label is functionally identical to this, just semantically different meaning.  In other words, I should be able to specify `service.kubernetes.io/service-proxy-name: not-kube-proxy` on a headless Service today, and get the same result.

Reply:
So, my understanding is that I can follow with this service.kubernetes.io/headless, as the idea is similar but the objective is different, right?

Reply:
yes, sorry to be confusing :)